noir_by_example/generic_traits/noir/src/main.nr

```diff
@@ -0,0 +1,84 @@
+// DCI (Data, Context, Interaction) example - keeps objects simple, and binds specific functions to them in their context of use
+
+// ### DATA ###
+struct ValueAccount {
+    balance: u32,
+}
+
+// Primitive functions for data object
+trait ValueFunctions {
+    fn add(&mut self, val: u32);
+    fn sub(&mut self, val: u32);
+}
+
+impl ValueFunctions for ValueAccount {
+    fn add(&mut self, val: u32) {
+        self.balance += val;
+    }
+    fn sub(&mut self, val: u32) {
+        self.balance -= val;
+    }
+}
+
+// ### CONTEXT ###
+
+// Example purchase context
+// Buyer role
+trait Buyer {
+    fn make_payment(&mut self, val: u32);
+}
+
+// impl <T:ValueFunctions> Buyer for T {   // This works in rust
+// impl <T> Buyer for T where T:ValueFunctions { // Valid noir, specifies value functions,
+//    but gives compilation error on line that calls Buyer function make_payment
+impl Buyer for ValueAccount { // context needs to be aware of data object (not just trait)
+    fn make_payment(&mut self, val: u32) {
+        self.sub(val);
+    }
+}
+
+// Seller role
+trait Seller {
+    fn receive_payment(&mut self, val: u32);
+}
+
+impl Seller for ValueAccount {
+    fn receive_payment(&mut self, val: u32) {
+        self.add(val);
+    }
+}
+
+// Define the purchase context with its generic roles
+struct PurchaseContext<B, S> {
+    // Roles
+    buyer: &mut B,
+    seller: &mut S,
+}
+
+impl <B, S> PurchaseContext<B, S> where B: Buyer, S: Seller {
+    fn do_purchase(&mut self, amount: u32) {
+        self.buyer.make_payment(amount); //   Compilation error "No matching impl found for `B: ValueFunctions`" when Buyer uses generic T
+        // self.buyer.receive_payment(amount); // correct compiler error, buyer role cannot receive_payment
+        self.seller.receive_payment(amount);
+    }
+}
+
+// ### INTERACTION ###
+
+fn main(start: u32, cost: pub u32) {
```

Comment:
`cargo build` is giving me this error (rustc 1.81.0 (eeb90cda1 2024-09-04))

![image](https://github.com/user-attachments/assets/edfd6c15-fffa-4daf-aba8-7f5abe8815d3)

Reply:
Noir gets params from Prover.toml, updated Rust to get its params from cmd line then call inner function. Test via direct call to inner func with params.